functions/export-csv.js

```diff
@@ -18,15 +18,84 @@
 'use strict';
 
 const {db} = require('./common/context');
+const CsvWriter = require('./common/csv-writer');
 
 function exportCsv(req, res) {
   const {
-    project: providedProjectId,
-    featureType: providedFeatureTypeId,
+    project: projectId,
+    featureType: featureTypeId,
     lang: desiredLanguage,
   } = req.query;
+  console.log('In the function');
 
-  return res.status(500).send('Not supported yet.');
+  let data = {};
+  let numberOfRecords = 0;
+  return db.fetchProject(projectId).then(
+    project => {
+      if (!project.exists) {
+        res.status(404).send('not found');
+        return Promise.reject(new Error('project not found: ' + projectId))
+      }
+      return Promise.all([
+        db.fetchForm(project.id, featureTypeId),
+        db.fetchRecords(project.id, featureTypeId)
+      ]);
+    }
+  ).then(
+    results => {
+      console.log(results);
+      const form = results[0];
+      for(var prop in form) {
+        if(form.hasOwnProperty(prop)) {
+          form[prop]['elements'].forEach(
+            field => {
+              data[field['id']] = [field['labels']['_']];
+            }
+          );
+        }
+      }
+
+      const records = results[1];
+      numberOfRecords = records.length;
+      records.forEach(
+        record => {
+          const responses = record.get('responses');
+          for(var prop in responses) {
+            toPush = response[prop];
+            if (Array.isArray(toPush)) {
+              toPush = '"' + toPush.join() + '"';
+            }
+            data[prop].push(toPush);
+          }
+        }
+      );
+      // let csvWriter = new CsvWriter(projectId, data, desiredLanguage ? desiredLanguage : '', numberOfRecords);
+      // return csvWriter.getTmpCsvFile();
+      return data;
+    }
+  ).then(
+    // tmpCsvFilePath => {
+    //   return res.download(tmpCsvFilePath);
+    // }
+    data => {
+      res.type('csv');
+      for (i = 0; i <= numberOfRecords; i++) {
+        toSend = '';
+        for (var prop in data) {
+          toSend += data[prop][i] + ',';
```

Comment:
It still appears you're building a giant string in memory. Perhaps you can just pass the res to the CsvWriter for now and let it call .write() on each line?